bundles/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/internal/NeeoRoomProtocol.java

```diff
@@ -123,23 +123,28 @@ public void processAction(NeeoAction action) {
     private void processScenarioChange(String scenarioKey, boolean launch) {
         NeeoUtil.requireNotEmpty(scenarioKey, "scenarioKey cannot be empty");
 
-        final List<String> activeScenarios = this.activeScenarios.get();
+        List<String> activeScenarios;
+        List<String> newActiveScenarios;
 
-        if (activeScenarios.contains(scenarioKey)) {
-            if (launch) {
-                return;
-            } else {
-                activeScenarios.remove(scenarioKey);
-            }
-        } else {
-            if (launch) {
-                activeScenarios.add(scenarioKey);
+        do {
+            activeScenarios = new ArrayList<>(this.activeScenarios.get());
+            newActiveScenarios = new ArrayList<>(activeScenarios);
+
+            if (activeScenarios.contains(scenarioKey)) {
+                if (launch) {
+                    return;
+                } else {
+                    newActiveScenarios.remove(scenarioKey);
+                }
             } else {
-                return;
+                if (launch) {
+                    newActiveScenarios.add(scenarioKey);
+                } else {
+                    return;
+                }
             }
-        }
+        } while (!this.activeScenarios.compareAndSet(activeScenarios, newActiveScenarios));
```

Comment:
This will never return true since you always make a copy of activeScenarios.get(). The compareAndSet does reference comparison.